examples/graphs/node2vec_movielens.py

```diff
@@ -0,0 +1,563 @@
+to learn movie embeddings. Such a dataset can be represented as a graph by treating
+the movies as nodes, and creating edges between movies that have similar ratings
+by the users. The learnt movie embeddings can be used for tasks such as movie recommendation,
+or movie genres prediction.
+
+This example requires `networkx` package, which can be installed using the following command:
+
+```python
+pip install networkx
+````
+"""
+
+"""
+## Setup
+"""
+
+import os
+from collections import defaultdict
+import math
+import networkx as nx
+import random
+from tqdm import tqdm
+from zipfile import ZipFile
+from urllib.request import urlretrieve
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+import matplotlib.pyplot as plt
+
+"""
+## Download and Prepare the Data
+
+The small version of the Movielens dataset includes around 100 thousand ratings
+from 610 users on 9742 movies.
+
+First, let's download the movielens data. The downloaded folder will contain
+three data files: `users.csv`, `movies.csv`, and `ratings.csv`. In this example,
+we will only need the `movies.dat`, and `ratings.dat` data files.
+"""
+
+urlretrieve(
+    "http://files.grouplens.org/datasets/movielens/ml-latest-small.zip", "movielens.zip"
+)
+ZipFile("movielens.zip", "r").extractall()
+
+"""
+Then, we load the data into pandas DataFrames and perform some basic processing.
+"""
+
+# Load movies to a DataFrame.
+movies = pd.read_csv("ml-latest-small/movies.csv")
+# Make the movieId string.
+movies["movieId"] = movies["movieId"].apply(lambda x: f"movie_{x}")
+
+# Load ratings to a DataFrame.
+ratings = pd.read_csv("ml-latest-small/ratings.csv")
+# # Convert rating to float
+ratings["rating"] = ratings["rating"].apply(lambda x: float(x))
+# # Make the movie_id string.
+ratings["movieId"] = ratings["movieId"].apply(lambda x: f"movie_{x}")
+
+print("Movies data shape:", movies.shape)
+print("Ratings data shape:", ratings.shape)
+
+"""
+Let's show a sample instance of the `ratings` DataFrame.
+"""
+
+ratings.head()
+
+"""
+Let's show how sample instance of the `movies` DataFrame.
+"""
+
+movies.head()
+
+"""
+Implement utility functions movies DataFrame
+"""
+
+
+def get_movie_title_by_id(movieId):
+    return list(movies[movies.movieId == movieId].title)[0]
+
+
+def get_movie_id_by_title(title):
+    return list(movies[movies.title == title].movieId)[0]
+
+
+"""
+## Construct the Movies Graph
+
+We create an edge between two movie nodes in the graph if both movies are rated
+by the same user >= `min_rating`. The weight of the edge will be based on the
+[pointwise mutual information](https://en.wikipedia.org/wiki/Pointwise_mutual_information)
+between the two movies, which is computed as: $log(xy) - log(x) - log(y) + log(D)$, where:
```

Comment:
We don't render latex, use backticks instead.

Reply:
Done